tool_path_planner/src/raster_path_generator.cpp

```diff
@@ -0,0 +1,337 @@
+ * to indicate whether the path should go A to B or B to A. A true value indicates A to B.
+ */
+struct SequencePoint
+{
+  size_t id;
+  bool from_a;
+};
+
+/**
+ * @brief From a sequence of path segments, this method extracts the end points and puts them into
+ * a new reference frame. As segments are indivisible, we only need the extremes for sorting them.
+ * @param segments The source of path segment data
+ * @param ref_rotation A transform from the origin to a reference frame which we want all the end
+ * points in. The paths in \e segments are considered to be in the origin frame.
+ * @return A sequence of end points in the reference frame of \e ref_rotation
+ */
+std::vector<PathEndPoints> toEndPoints(const std::vector<EigenSTL::vector_Affine3d>& segments,
+                                       const Eigen::Quaterniond& ref_rotation)
+{
+  // Ref rotation is the Target Frame w.r.t. Origin
+  // The points are all w.r.t. Origin, ergo we have to pre-multiply by the inverse of ref_rotation
+  // to get the new points in the Target Frame
+  Eigen::Affine3d ref_inv;
+  ref_inv = ref_rotation.inverse();
+
+  std::vector<PathEndPoints> result;
+  for (std::size_t i = 0; i < segments.size(); ++i)
+  {
+    const auto& s = segments[i];
+    Eigen::Vector3d a = (ref_inv * s.front()).translation();
+    Eigen::Vector3d b = (ref_inv * s.back()).translation();
+    result.push_back({a, b, i});
+  }
+  return result;
+}
+
+/**
+ * @brief Reconstructs a set of PoseArray objects using the given set of sequence points which contain
+ * indices into the \e end_points array which reference the original \e in trajectory.
+ * @param in The original trajectory
+ * @param seqs The sequence points whose 'id' field reaches into the \e end_points vector
+ * @param end_points The set of end points whose 'id' field reaches into the \e in vector
+ * @return A new pose array constructed with the sequence ordering from the \e in trajectory
+ */
+std::vector<geometry_msgs::PoseArray> makeSequence(const std::vector<geometry_msgs::PoseArray>& in,
+                                                   const std::vector<SequencePoint>& seqs,
+                                                   const std::vector<PathEndPoints>& end_points)
+{
+  assert(in.size() == seqs.size());
+  std::vector<geometry_msgs::PoseArray> rs;
+  rs.reserve(in.size());
+
+  for (const auto& seq : seqs)
+  {
+    rs.push_back(in[end_points[seq.id].id]); // seq.id points to end_points; end_points.id points to in
+    if (!seq.from_a) // The 'in' trajectory has segments that are always A to B
+    {
+      std::reverse(rs.back().poses.begin(), rs.back().poses.end());
+    }
+  }
+
+  return rs;
+}
+
+/**
+ * @brief Computes the 'average' quaternion from an input set of them.
+ * See http://stackoverflow.com/questions/12374087/average-of-multiple-quaternions
+ * See http://www.acsu.buffalo.edu/~johnc/ave_quat07.pdf
+ *
```

Comment:
As we discussed offline, yes the averaging of the orientations expressed in the angle-axis form (3d vector) could work as long as the vectors are within the same neighborhood.  However I'm not sure that way of doing it will have significant advantages over the quaternion approach implemented here.